File: BetterTridents/src/main/java/project/kompass/btk/listener/CopperArmorListener.java
package project.kompass.btk.listener;

import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.entity.Entity;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Mob;
import org.bukkit.entity.Player;
import org.bukkit.event.Listener;
import org.bukkit.inventory.EntityEquipment;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.HashSet;
import java.util.Random;
import java.util.Set;

public class CopperArmorListener implements Listener {

    private final Random random = new Random();

    // 1. Core strike check task (Executes every 30 seconds)
    public void startArmorCheckTask(JavaPlugin plugin) {
        new BukkitRunnable() {
            @Override
            public void run() {
                Set<LivingEntity> processed = new HashSet<>();

                for (Player player : Bukkit.getOnlinePlayers()) {
                    World world = player.getWorld();
                    if (!world.hasStorm()) continue;

                    int playerHighestY = world.getHighestBlockYAt(player.getLocation());
                    if (player.getLocation().getBlockY() < playerHighestY) continue;

                    // Process player
                    if (!processed.contains(player)) {
                        processed.add(player);
                        checkAndStrike(player, world);
                    }

                    // Process nearby mobs within a 32-block radius
                    double radius = 32.0;
                    for (Entity nearby : player.getNearbyEntities(radius, radius, radius)) {
                        if (nearby instanceof Mob mob) {
                            if (processed.contains(mob)) continue;
                            processed.add(mob);

                            int mobHighestY = world.getHighestBlockYAt(mob.getLocation());
                            if (mob.getLocation().getBlockY() < mobHighestY) continue;

                            checkAndStrike(mob, world);
                        }
                    }
                }
            }
        }.runTaskTimer(plugin, 600L, 600L);
    }

    // 2. High-performance ambient particle task (Executes every 10 ticks / 0.5 seconds)
    public void startParticleTask(JavaPlugin plugin) {
        new BukkitRunnable() {
            @Override
            public void run() {
                Set<LivingEntity> processed = new HashSet<>();

                for (Player player : Bukkit.getOnlinePlayers()) {
                    World world = player.getWorld();
                    if (!world.hasStorm()) continue;

                    // Process player particles
                    if (!processed.contains(player)) {
                        processed.add(player);
                        int playerHighestY = world.getHighestBlockYAt(player.getLocation());
                        if (player.getLocation().getBlockY() >= playerHighestY) {
                            spawnCopperParticlesIfEligible(player, world);
                        }
                    }

                    // Process nearby mob particles within a 32-block radius
                    double radius = 32.0;
                    for (Entity nearby : player.getNearbyEntities(radius, radius, radius)) {
                        if (nearby instanceof Mob mob) {
                            if (processed.contains(mob)) continue;
                            processed.add(mob);

                            int mobHighestY = world.getHighestBlockYAt(mob.getLocation());
                            if (mob.getLocation().getBlockY() >= mobHighestY) {
                                spawnCopperParticlesIfEligible(mob, world);
                            }
                        }
                    }
                }
            }
        }.runTaskTimer(plugin, 10L, 10L);
    }

    private void checkAndStrike(LivingEntity entity, World world) {
        EntityEquipment equipment = entity.getEquipment();
        if (equipment == null) return;

        ItemStack[] armorContents = equipment.getArmorContents();
        int copperCount = 0;
        for (ItemStack item : armorContents) {
            if (item != null && isCopperArmor(item.getType())) {
                copperCount++;
            }
        }
        if (copperCount == 0) return;

        double chance = copperCount * 0.25;
        if (random.nextDouble() < chance) {
            world.strikeLightning(entity.getLocation());
        }
    }

    private void spawnCopperParticlesIfEligible(LivingEntity entity, World world) {
        EntityEquipment equipment = entity.getEquipment();
        if (equipment == null) return;

        ItemStack[] armorContents = equipment.getArmorContents();
        int copperCount = 0;
        for (ItemStack item : armorContents) {
            if (item != null && isCopperArmor(item.getType())) {
                copperCount++;
            }
        }
        if (copperCount == 0) return;

        // Scale the particle quantity based on the piece count
        int particleCount = copperCount * 2;
        if (copperCount == 4) {
            particleCount = 10; // Extra intense for full sets
        }

        // Adjust boundaries dynamically to scale the "leap" width of the sparks
        double offsetX = 0.15 * copperCount;
        double offsetY = entity.getHeight() / 2.0;
        double offsetZ = 0.15 * copperCount;

        // Spawn ELECTRIC_SPARK particles relative to the entity's center point
        world.spawnParticle(
                org.bukkit.Particle.ELECTRIC_SPARK,
                entity.getLocation().add(0, offsetY, 0),
                particleCount,
                offsetX,
                offsetY,
                offsetZ,
                0.01 // Minimal particle speed/noise parameter
        );
    }

    private boolean isCopperArmor(Material type) {
        return type == Material.COPPER_HELMET ||
                type == Material.COPPER_CHESTPLATE ||
                type == Material.COPPER_LEGGINGS ||
                type == Material.COPPER_BOOTS;
    }
}

File: BetterTridents/src/main/kotlin/project/kompass/btk/listener/CopperArmorListener.kt
package project.kompass.btk.listener

import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.Particle
import org.bukkit.World
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Mob
import org.bukkit.event.Listener
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.scheduler.BukkitRunnable
import java.util.EnumSet
import java.util.HashSet
import java.util.Random

class CopperArmorListener : Listener {

    private val random = Random()

    companion object {
        private val COPPER_ARMOR_MATERIALS: Set<Material> = EnumSet.of(
            Material.COPPER_HELMET,
            Material.COPPER_CHESTPLATE,
            Material.COPPER_LEGGINGS,
            Material.COPPER_BOOTS
        )
    }

    fun startArmorCheckTask(plugin: JavaPlugin) {
        object : BukkitRunnable() {
            override fun run() {
                val processed = HashSet<LivingEntity>()

                for (player in Bukkit.getOnlinePlayers()) {
                    val world = player.world
                    if (!world.hasStorm()) continue

                    val playerHighestY = world.getHighestBlockYAt(player.location)
                    if (player.location.blockY < playerHighestY) continue

                    if (!processed.contains(player)) {
                        processed.add(player)
                        checkAndStrike(player, world)
                    }

                    val radius = 32.0
                    for (nearby in player.getNearbyEntities(radius, radius, radius)) {
                        if (nearby is Mob) {
                            if (processed.contains(nearby)) continue
                            processed.add(nearby)

                            val mobHighestY = world.getHighestBlockYAt(nearby.location)
                            if (nearby.location.blockY < mobHighestY) continue

                            checkAndStrike(nearby, world)
                        }
                    }
                }
            }
        }.runTaskTimer(plugin, 600L, 600L) // Runs every 30 seconds
    }

    fun startParticleTask(plugin: JavaPlugin) {
        object : BukkitRunnable() {
            override fun run() {
                val processed = HashSet<LivingEntity>()

                for (player in Bukkit.getOnlinePlayers()) {
                    val world = player.world
                    if (!world.hasStorm()) continue

                    if (!processed.contains(player)) {
                        processed.add(player)
                        val playerHighestY = world.getHighestBlockYAt(player.location)
                        if (player.location.blockY >= playerHighestY) {
                            spawnCopperParticlesIfEligible(player, world)
                        }
                    }

                    val radius = 32.0
                    for (nearby in player.getNearbyEntities(radius, radius, radius)) {
                        if (nearby is Mob) {
                            if (processed.contains(nearby)) continue
                            processed.add(nearby)

                            val mobHighestY = world.getHighestBlockYAt(nearby.location)
                            if (nearby.location.blockY >= mobHighestY) {
                                spawnCopperParticlesIfEligible(nearby, world)
                            }
                        }
                    }
                }
            }
        }.runTaskTimer(plugin, 10L, 10L) // Runs every 0.5 seconds
    }

    private fun checkAndStrike(entity: LivingEntity, world: World) {
        val equipment = entity.equipment ?: return

        val armorContents = equipment.armorContents
        var copperCount = 0
        for (item in armorContents) {
            if (item != null && isCopperArmor(item.type)) {
                copperCount++
            }
        }
        if (copperCount == 0) return

        val chance = copperCount * 0.25
        if (random.nextDouble() < chance) {
            world.strikeLightning(entity.location)
        }
    }

    private fun spawnCopperParticlesIfEligible(entity: LivingEntity, world: World) {
        val equipment = entity.equipment ?: return

        val armorContents = equipment.armorContents
        var copperCount = 0
        for (item in armorContents) {
            if (item != null && isCopperArmor(item.type)) {
                copperCount++
            }
        }
        if (copperCount == 0) return

        var particleCount = copperCount * 2
        if (copperCount == 4) {
            particleCount = 10
        }

        val offsetX = 0.15 * copperCount
        val offsetY = entity.height / 2.0
        val offsetZ = 0.15 * copperCount

        world.spawnParticle(
            Particle.ELECTRIC_SPARK,
            entity.location.add(0.0, offsetY, 0.0),
            particleCount,
            offsetX,
            offsetY,
            offsetZ,
            0.01
        )
    }

    private fun isCopperArmor(type: Material): Boolean {
        return COPPER_ARMOR_MATERIALS.contains(type)
    }
}